### ISSUE TITLE
work unnormally when tracing nested spans

### ISSUE THREAD

**Issue by nullscc** [Yelp/py_zipkin#96]
I have encountered a problem when I use py_zipkin which version is 0.13.0. But first, let me show you the sample code I use.

## sample code
first, the helper function for `transport_handler` in a file named `handler.py`
```
# handler.py

def default_handler(encoded_span):
    # for version 0.7
    # body = str.encode('\x0c\x00\x00\x00\x01') + encoded_span

    # for version 0.13
    body = encoded_span

    return requests.post(
        "http://127.0.0.1:9411/api/v1/spans",
        data=body,
        headers={'Content-Type': 'application/x-thrift'},
    )
```

and then, the file `demo.py` which is a flask application offer api to client or browser, which export port 5005
```
# demo.py

import requests
from flask import Flask
from py_zipkin.zipkin import zipkin_span,create_http_headers_for_new_span
import time
from handler import default_handler as http_transport

app = Flask(__name__)

@zipkin_span(service_name='webapp', span_name='do_stuff')
def do_stuff():
    headers = create_http_headers_for_new_span()
    requests.get('http://localhost:6000/service1/', headers=headers)
    return 'OK'

@app.route('/')
def index():
    print('recieve request in demo', int(time.time()))
    with zipkin_span(
        service_name='webapp',
        span_name='index',
        transport_handler=http_transport,
        port=5005,
        sample_rate=100, #0.05, # Value between 0.0 and 100.0
    ) as con:
        do_stuff()
    return 'OK', 200

if __name__=='__main__':
    app.run(host="0.0.0.0",port=5005,debug=True)
```

and last but not least, the file named `service1.py` which is another flask application offer an api for `demo.py` to mock a microservice, which expose port 6000 
```
# service1.py
import requests
from flask import Flask, request
from py_zipkin.zipkin import zipkin_span,create_http_headers_for_new_span, ZipkinAttrs
import time
from handler import default_handler as http_transport

app = Flask(__name__)

@zipkin_span(service_name='service1', span_name='service1_do_stuff')
def do_stuff():
    return 'OK'

@app.route('/service1/')
def index():
    with zipkin_span(
        service_name='service1',
        zipkin_attrs=ZipkinAttrs(
            trace_id=request.headers['X-B3-TraceID'],
            span_id=request.headers['X-B3-SpanID'],
            parent_span_id=request.headers['X-B3-ParentSpanID'],
            flags=request.headers['X-B3-Flags'],
            is_sampled=request.headers['X-B3-Sampled'],
        ),
        span_name='index_service1',
        transport_handler=http_transport,
        sample_rate=100, #0.05, # Value between 0.0 and 100.0
    ):
        do_stuff()
    return 'OK', 200

if __name__=='__main__':
    app.run(host="0.0.0.0",port=6000,debug=True)
```

## problem
I access `http://127.0.0.1:5005`,  and  there are no error complaints in `demo.py` and `service1.py`, Here is the output when I browse `http://127.0.0.1:9411` which sort option is `Newest first`.  :
<img width="1354" alt="screen shot 2018-08-21 at 11 20 38 am" src="https://user-images.githubusercontent.com/17700282/44378519-489e0c80-a534-11e8-946e-618c94156f13.png">
so the first problem is no nested span(`do_stuff` and `service1_do_stuff`) showed there, the other is that there is no child-parent relationship showed there between `webapp` and `service1`.

## experiment
But when I try to change the py_zipkin's version to 0.7 and just changed this line ( function `default_handler` in `handler.py`):
```
    # for version 0.13
    body = encoded_span
```
to this:
```
    # for version 0.7
    body = str.encode('\x0c\x00\x00\x00\x01') + encoded_span
```
and no other code changed any more.
"weird" thing happened(Newest first of course)
<img width="1348" alt="screen shot 2018-08-21 at 11 29 34 am" src="https://user-images.githubusercontent.com/17700282/44378789-8485a180-a535-11e8-85cd-0440cf7dfa66.png">

I don't know whether the usage is right or not?or there are some bugs in newest version for nested spans within in a microservice?
Any help or reply will be greatly appreciated!

**Comment by drolando** [comment]
Hmm weird. Thanks for the detailed explanation and the code sample! I'll try to reproduce it locally and debug what's going on.

**Comment by nullscc** [comment]
@drolando 
Oh, thank you. We intend to use it in our production environment.

**Comment by drolando** [comment]
@nullscc btw, what version of zipkin-server are you using?

**Comment by nullscc** [comment]
@drolando Oh, sorry for not providing that kind of information.
I just run this command 2-3 days ago.
```
docker run -d -p 9411:9411 openzipkin/zipkin
```
I think it just the latest version.

**Comment by drolando** [comment]
@nullscc It looks like using `zipkin_span` as a decorator is broken. If you do:
```python
with zipkin_span(
    service_name='webapp',
    span_name='index',
    transport_handler=http_transport,
    port=5005,
    sample_rate=100, #0.05, # Value between 0.0 and 100.0
) as con:
    with zipkin_span(service_name='webapp', span_name='do_stuff_1'):
        do_stuff()
```
then `do_stuff_1` shows up. Afaik we almost never use it as a decorator, which explains why we didn't notice it. It should be easy to fix

**Comment by nullscc** [comment]
oh, It's nice, I'll try it. And I have notice another problem: when I install py_zipkin using pip, the version is 0.1? I am not sure about it though...

**Comment by drolando** [comment]
Weird, https://pypi.org/project/py_zipkin/ has the latest version.

**Comment by nullscc** [comment]
My mistake, I'm sorry about that. Maybe my eye is blur...
@drolando

**Comment by nullscc** [comment]
@drolando You are right, when I change the code to this:
```
def zipkin_span_decorate(*args_out, **kw_out):
    def decorate(f):
        def inner(*args, **kw):
            with zipkin_span(*args_out, **kw_out):
                return f(*args, **kw)
        return inner
    return decorate

@zipkin_span_decorate(service_name='webapp', span_name='do_stuff')
def do_stuff():
    headers = create_http_headers_for_new_span()
    requests.get('http://localhost:6000/service1/', headers=headers)
    return 'OK'
```
The thing goes normally, thank you so much. 

You know I also looked into the source code of py_zipkin, and just see there's no different between decorator and context_manager. So I never mind this problem are caused by decorator.

**Comment by nullscc** [comment]
@drolando Oh, I get it.
We could use it only in the form of context manager. But the way I show in last comment still can let us use a form of decorator, because the code `zipkin_span ` still run in runtime.

**Comment by drolando** [comment]
@nullscc unrelated to this, but please use the new class-based transport handlers as shown at https://github.com/Yelp/py_zipkin#transport

Using a single function transport handler is deprecated.

**Comment by nullscc** [comment]
@drolando  Ok, I nerver realize it, thanks for reminding me.

**Comment by bplotnick** [comment]
For the thead-local case, I think having a new decorator function that defers the creation of the span object until the decorated function is called would be fine (which is what @nullscc's example is doing).

The non-thread-local case is a little bit trickier. One idea would be to use reflection to pass the context as arguments to the decorated function:

```python
@zipkin_decorator(service_name='some_service', span_name='some_span', context_arg='context')
def foo(context, other, args):
   ...
```

The implementation would use inspect - https://stackoverflow.com/a/16542145

It seems like it might be a bit messy, but possible. I'll try to think of other ways to do this.

**Comment by nullscc** [comment]
@bplotnick could you tell me what are thead-local and non-thread-local case? 
And if I want to use in a multi-threading environment, these is something special to do?

**Comment by bplotnick** [comment]
@nullscc By "thread-local case" i mean single-threaded without [cooperative multitasking|https://github.com/Yelp/py_zipkin#using-in-multithreading-evironments], e.g. asyncio. If you are in this situation, you can just let py_zipkin propagate context for you via thread-local.

If you are using either multiple hardware threads or multiple green threads or whatever, you need to explicitly pass context. This is why we have the `context_stack` override, and you'll need to use it if you are in a multithreaded environment.

So what i was proposing was that we can have the decorator inject the context into the argument to make it a bit easier to use. btw, I totally ripped this idea off of go: https://blog.golang.org/context

**Comment by nullscc** [comment]
@bplotnick I'm sorry, that seems a little tricky to me. This may due to my understanding about the concept of you description of thread. I've been see that you derive three concepts:
* single-threaded without cooperative multitasking
* multiple hardware threads
* multiple green threads

And I have list some example to reflect those concepts, could you help me confirm it?
* tornado : single-threaded without cooperative multitasking
* flask : multiple hardware threads
* gevent : multiple green threads

To be more precisely about my question, if I want to use it in flask or a typical wsgi application which is multiple-threading evironment, should I explicitly pass context? If I should pass it, that symbolizes my example code showed above have some mistake. And that if I use it in tornado which just like asyncio?

**Comment by nullscc** [comment]
@zhuwei05 could you show me the code? You just copy the code from this issue or from my repos?

**Comment by zhuwei05** [comment]
@nullscc  I just copy all you code with `zipkin_span_decorate`

**Comment by nullscc** [comment]
@zhuwei05 emmm.. werid, maybe caused by network? It happens every time you execute it?

**Comment by drolando** [comment]
Fixed by #118
